File: backend/internal/model/enrollment.go
package model

import "time"

// Enrollment запись студента на курс
type Enrollment struct {
	ID         int64     `json:"id"`
	UserID     int64     `json:"user_id"`
	CourseID   int64     `json:"course_id"`
	EnrolledAt time.Time `json:"enrolled_at"`
	Course     *Course   `json:"course,omitempty"`
}

// LessonProgress прогресс студента по уроку
type LessonProgress struct {
	ID          int64      `json:"id"`
	UserID      int64      `json:"user_id"`
	LessonID    int64      `json:"lesson_id"`
	QuizScore   *int       `json:"quiz_score"`
	CompletedAt *time.Time `json:"completed_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// EnrollmentListResponse список записей с курсами
type EnrollmentListResponse struct {
	Items []Enrollment `json:"items"`
	Total int          `json:"total"`
}

// CourseProgressSummary сводка прогресса по курсу
type CourseProgressSummary struct {
	CourseID       int64   `json:"course_id"`
	CourseTitle    string  `json:"course_title"`
	CourseSlug     string  `json:"course_slug"`
	TotalLessons   int     `json:"total_lessons"`
	CompletedCount int     `json:"completed_count"`
	Percent        int     `json:"percent"`
	LastLessonSlug string  `json:"last_lesson_slug,omitempty"`
	LastLessonTitle string `json:"last_lesson_title,omitempty"`
}

File: frontend/lib/api/public/enrollment.ts
import { apiGet, apiPost } from '../client';
import type { Course } from '../types';

export interface Enrollment {
  id: number;
  user_id: number;
  course_id: number;
  enrolled_at: string;
  course?: Course;
}

export interface CourseProgressSummary {
  course_id: number;
  course_title: string;
  course_slug: string;
  total_lessons: number;
  completed_count: number;
  percent: number;
  last_lesson_slug?: string;
  last_lesson_title?: string;
}

export interface LessonProgressState {
  id?: number;
  user_id?: number;
  lesson_id?: number;
  quiz_score?: number | null;
  completed_at?: string | null;
  completed: boolean;
}

// === Enrollments ===

export async function enrollInCourse(slug: string): Promise<Enrollment> {
  const res = await apiPost<{ enrollment: Enrollment; message: string }>(
    `/api/v1/courses/${slug}/enroll`,
    {}
  );
  return res.enrollment;
}

export async function getMyEnrollments(): Promise<{ items: Enrollment[]; total: number }> {
  return apiGet<{ items: Enrollment[]; total: number }>('/api/v1/me/enrollments');
}

export async function getEnrollmentStatus(courseId: number): Promise<{ enrolled: boolean }> {
  return apiGet<{ enrolled: boolean }>(`/api/v1/courses/${courseId}/enrollment-status`);
}

// === Progress ===

export async function markLessonComplete(lessonId: number): Promise<LessonProgressState> {
  return apiPost<LessonProgressState>(`/api/v1/lessons/${lessonId}/complete`, {});
}

export async function submitQuiz(
  lessonId: number,
  score: number
): Promise<LessonProgressState> {
  return apiPost<LessonProgressState>(`/api/v1/lessons/${lessonId}/quiz`, { score });
}

export async function getCourseProgress(courseId: number): Promise<CourseProgressSummary> {
  return apiGet<CourseProgressSummary>(`/api/v1/me/courses/${courseId}/progress`);
}

export async function getLessonProgress(lessonId: number): Promise<LessonProgressState> {
  return apiGet<LessonProgressState>(`/api/v1/lessons/${lessonId}/progress`);
}

export interface LessonProgressItem {
  id: number;
  user_id: number;
  lesson_id: number;
  quiz_score: number | null;
  completed_at: string | null;
  updated_at: string;
}

export async function getMyLessonsProgress(
  courseId: number
): Promise<{ items: LessonProgressItem[] }> {
  return apiGet<{ items: LessonProgressItem[] }>(
    `/api/v1/me/courses/${courseId}/lessons-progress`
  );
}
